# Agricultural Extension RAG - Smart Retrieval for Farmers (v7: Ultimate)
### Team Gambia | TRI AI Saturdays Cohort 10

> **This is the notebook that produced our final competition submission.** Local nDCG@5 on the 308 training queries: **0.782** (5-fold CV: 0.782 ± 0.024). Actual competition result once the private test set was scored: **Public leaderboard 0.83823, Private leaderboard 0.86982** — 8th place overall. As discussed in Section 15 below, the private score being *higher* than both the public score and our local estimate is a recurring pattern in this project: local training-set nDCG@5 has been a directionally useful but imperfect predictor of true held-out performance throughout development.

## What this notebook does

Builds a retrieval engine that ranks agricultural extension factsheets against a farmer's natural-language question, scored by **nDCG@5**. nDCG rewards both finding the right document and ranking it near the top -- a correct document at rank 5 contributes roughly 40% less credit than the same document at rank 1, purely from position.

## Why this problem is harder than plain keyword search

The corpus is deliberately built to punish naive approaches: farmers phrase questions colloquially ("prevent") while factsheets use formal titles ("Preventing..."), and the labels include **hard negatives** -- documents that share vocabulary with a query but answer a *different, adjacent* problem (e.g. potassium deficiency retrieved for a nitrogen-deficiency question), specifically to penalize systems that rely on lexical overlap alone.

## What this notebook resolves, and why it's structured as a set of experiments rather than a fixed pipeline

Earlier iterations (v4-v6) added several plausible-sounding improvements -- lemmatization instead of stemming, a larger reranker, query expansion -- all at once, and the result was a *worse* leaderboard score with no way to tell which change caused it. The lesson: every design choice below is tested in isolation against real labeled data (`qrels_train.csv`) before being adopted, rather than assumed to help because it sounds architecturally reasonable.

**Two specific open questions this notebook resolves:**

1. **Did lemmatization/expanded queries quietly hurt dense retrieval?** An earlier dense-only score (0.7688) had dropped to 0.7495 after those two changes were introduced together, and neither had been tested in isolation. This notebook isolates title-doubling (does repeating the title in the document text help a bi-encoder?) and query expansion (does adding scientific-jargon synonyms help or hurt a semantic embedding query?) separately, so the actual cause can be identified rather than guessed.
2. **Stemming or lemmatization for the lexical (BM25/TF-IDF) side?** Never tested head-to-head in the same run before -- this notebook does, crossed with raw-vs-expanded query, so the winner is chosen on evidence.

**Carried forward as already-proven, not re-litigated here:**
- Porter stemming (in whichever morphology wins below) fixes a real, diagnosed failure mode: template document clusters differing by a single intent word (e.g. "Preventing X" vs "Managing X"), where an unstemmed query like "prevent X" never lexically matches the correct document's title word "Preventing".
- Convex combination fusion beats Reciprocal Rank Fusion (RRF), because RRF only sees rank position and discards how *confident* each retriever actually was -- measured directly in an earlier run, RRF fusion (0.640) was worse than dense retrieval alone (0.769).
- `ms-marco-MiniLM-L-6-v2` beats a larger cross-encoder (`bge-reranker-base`, ~12x more parameters) here, zero-shot, at a fraction of the compute -- confirmed directly in an earlier run, not assumed from model size.

**New this round:** after the ablations pick a final configuration, that fixed configuration gets a **5-fold cross-validation** pass -- not to search for better hyperparameters, but to report an honest mean +/- standard deviation instead of a single number that risks being an artifact of tuning repeatedly against the same 308 training queries.

In [1]:
!pip install -q rank_bm25 sentence-transformers nltk


In [2]:
import os, re, glob, nltk
import numpy as np, pandas as pd
from tqdm.auto import tqdm
from collections import defaultdict
from rank_bm25 import BM25Okapi
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.metrics.pairwise import cosine_similarity
from sklearn.model_selection import KFold
from sentence_transformers import SentenceTransformer, CrossEncoder
from nltk.stem import PorterStemmer, WordNetLemmatizer
from nltk.corpus import wordnet
from nltk.tokenize import word_tokenize

pd.set_option("display.max_colwidth", 120)
np.random.seed(42)

nltk.download('punkt', quiet=True)
nltk.download('punkt_tab', quiet=True)
nltk.download('averaged_perceptron_tagger', quiet=True)
nltk.download('averaged_perceptron_tagger_eng', quiet=True)
nltk.download('wordnet', quiet=True)
nltk.download('omw-1.4', quiet=True)
stemmer = PorterStemmer()
lemmatizer = WordNetLemmatizer()


## Locate and load the competition data
We search for `documents.csv` under `/kaggle/input/` rather than hard-coding a path, since Kaggle's auto-generated input slug doesn't always exactly match the competition's URL slug -- this keeps the notebook working without manual edits if that mismatch occurs.

In [3]:
candidates = glob.glob("/kaggle/input/**/documents.csv", recursive=True)
if not candidates:
    raise FileNotFoundError("documents.csv not found under /kaggle/input -- attach the competition data in the Input panel.")
DATA_DIR = os.path.dirname(candidates[0])
print("Using data directory:", DATA_DIR)

docs_df = pd.read_csv(f"{DATA_DIR}/documents.csv")
train_queries_df = pd.read_csv(f"{DATA_DIR}/train_queries.csv")
qrels_train_df = pd.read_csv(f"{DATA_DIR}/qrels_train.csv")
test_queries_df = pd.read_csv(f"{DATA_DIR}/test_queries.csv")

DOC_ID_COL, DOC_TITLE_COL, DOC_TEXT_COL = "document_id", "title", "text"
QUERY_ID_COL, QUERY_TEXT_COL = "query_id", "query"
REL_QUERY_ID_COL, REL_DOC_ID_COL, RELEVANCE_COL = "query_id", "document_id", "relevance"

for col in [DOC_ID_COL, DOC_TITLE_COL, DOC_TEXT_COL]:
    assert col in docs_df.columns, f"'{col}' missing from documents.csv."
for col in [QUERY_ID_COL, QUERY_TEXT_COL]:
    assert col in test_queries_df.columns, f"'{col}' missing from test_queries.csv."
for col in [REL_QUERY_ID_COL, REL_DOC_ID_COL, RELEVANCE_COL]:
    assert col in qrels_train_df.columns, f"'{col}' missing from qrels_train.csv."

print(f"Documents: {len(docs_df)}  Train queries: {len(train_queries_df)}  Qrels: {len(qrels_train_df)}  Test queries: {len(test_queries_df)}")


Using data directory: /kaggle/input/competitions/agricultural-extension-rag-smart-retrieval-for-farmers
Documents: 695  Train queries: 308  Qrels: 4194  Test queries: 200


## Query expansion (tested in isolation on dense AND lexical separately below)

**Intuition:** farmers may name a crop problem using a local/colloquial term (e.g. "matooke wilt") while the factsheets use scientific nomenclature ("Banana Bacterial Wilt / *Xanthomonas campestris*"). Bag-of-words methods like BM25/TF-IDF have no way to connect these unless the vocabulary literally overlaps, so we bridge the gap by appending the scientific terms whenever a query mentions a recognized colloquial trigger.

We do **not** assume this transfers to dense (semantic) retrieval -- a bi-encoder already generalizes across paraphrasing on its own, so appending extra jargon could just as easily add noise to the sentence it's embedding rather than help. Ablation 2 below tests this directly instead of assuming either outcome.

In [4]:
def expand_local_agri_queries(query_text):
    if not isinstance(query_text, str):
        return query_text
    q = query_text.lower()
    expansions = []
    if "matooke" in q or "banana" in q:
        if "wilt" in q or "rot" in q:
            expansions.append("bxw banana bacterial wilt xanthomonas")
    if "cassava" in q and ("rot" in q or "streak" in q):
        expansions.append("cbsd cassava brown streak disease")
    if "bean" in q:
        if "beetle" in q or "eating" in q:
            expansions.append("ootheca leaf beetle")
        if "spot" in q or "rust" in q:
            expansions.append("anthracnose colletotrichum")
    if "maize" in q and ("worm" in q or "borer" in q):
        expansions.append("faw fall armyworm spodoptera frugiperda")
    if ("cereal" in q or "maize" in q) and "weed" in q:
        expansions.append("striga")
    if "coffee" in q and ("borer" in q or "twig" in q):
        expansions.append("black coffee twig borer xylosandrus compactus cwd")
    if expansions:
        return re.sub(r"\s+", " ", query_text + " " + " ".join(expansions)).strip()
    return query_text

train_queries_df["expanded_query"] = train_queries_df[QUERY_TEXT_COL].apply(expand_local_agri_queries)
test_queries_df["expanded_query"] = test_queries_df[QUERY_TEXT_COL].apply(expand_local_agri_queries)
n_expanded = (train_queries_df["expanded_query"] != train_queries_df[QUERY_TEXT_COL]).sum()
print(f"Train queries modified by expansion: {n_expanded} / {len(train_queries_df)}")


Train queries modified by expansion: 9 / 308


## Preprocessing: build all text variants needed for the ablations below
- **Lexical doc text** (BM25/TF-IDF): title-doubled (a deliberate term-frequency boost -- BM25/TF-IDF score based on term counts, so repeating the title is a lightweight way to weight a document's most concise topic indicator more heavily). Built two ways: **stemmed** and **lemmatized**, so Ablation 3 below can decide between them on evidence rather than preference.
- **Dense doc text**: built TWO ways -- title-single and title-doubled -- so the doubling's effect on a bi-encoder can actually be measured (Ablation 1) instead of assumed to transfer from the lexical pipeline, where the mechanism for *why* it helps is completely different (term-frequency counting vs. a single pooled vector).
- **Query text**: raw and expanded, crossed with stem/lemma for lexical, and left as plain cleaned text for dense (no stemming/lemmatization there -- a transformer's own subword tokenizer already generalizes across "prevent"/"preventing" without help, so stemming that text would only remove information).

In [5]:
def get_wordnet_pos(treebank_tag):
    if treebank_tag.startswith('J'): return wordnet.ADJ
    if treebank_tag.startswith('V'): return wordnet.VERB
    if treebank_tag.startswith('N'): return wordnet.NOUN
    if treebank_tag.startswith('R'): return wordnet.ADV
    return wordnet.NOUN

def clean_and_process(text, mode="none"):
    if pd.isna(text):
        return ""
    text = str(text).lower()
    text = re.sub(r"[\r\n\t]+", " ", text)
    text = re.sub(r"[^a-z0-9\s\-%.]", " ", text)
    text = re.sub(r"\s+", " ", text).strip()
    if mode == "stem":
        text = " ".join(stemmer.stem(w) for w in text.split())
    elif mode == "lemma":
        tokens = word_tokenize(text)
        pos_tags = nltk.pos_tag(tokens)
        text = " ".join(lemmatizer.lemmatize(w, pos=get_wordnet_pos(t)) for w, t in pos_tags)
    return text

def build_doc_text(row, double_title):
    title = str(row[DOC_TITLE_COL]) if pd.notna(row.get(DOC_TITLE_COL, None)) else ""
    if double_title:
        prefix = f"{title}. {title}. " if title else ""
    else:
        prefix = f"{title}. " if title else ""
    return prefix + str(row[DOC_TEXT_COL])

doc_ids = docs_df[DOC_ID_COL].tolist()

# Lexical doc text (title-doubled, as before) x {stem, lemma}
docs_df["lex_raw"] = docs_df.apply(lambda r: build_doc_text(r, double_title=True), axis=1)
lex_doc_texts = {mode: docs_df["lex_raw"].apply(lambda t: clean_and_process(t, mode)).tolist() for mode in ["stem", "lemma"]}

# Dense doc text: title-single vs title-doubled, both unstemmed/unlemmatized
docs_df["dense_raw_single"] = docs_df.apply(lambda r: build_doc_text(r, double_title=False), axis=1)
docs_df["dense_raw_double"] = docs_df.apply(lambda r: build_doc_text(r, double_title=True), axis=1)
dense_doc_texts = {
    "single": docs_df["dense_raw_single"].apply(lambda t: clean_and_process(t, "none")).tolist(),
    "double": docs_df["dense_raw_double"].apply(lambda t: clean_and_process(t, "none")).tolist(),
}

# Query text variants
for df in (train_queries_df, test_queries_df):
    for mode in ["stem", "lemma"]:
        df[f"q_lex_{mode}_raw"] = df[QUERY_TEXT_COL].apply(lambda t: clean_and_process(t, mode))
        df[f"q_lex_{mode}_expanded"] = df["expanded_query"].apply(lambda t: clean_and_process(t, mode))
    df["q_dense_raw"] = df[QUERY_TEXT_COL].apply(lambda t: clean_and_process(t, "none"))
    df["q_dense_expanded"] = df["expanded_query"].apply(lambda t: clean_and_process(t, "none"))

print("All text variants built.")


All text variants built.


## Ground truth (qrels) + nDCG@5 evaluation harness

**The metric, intuitively:** $DCG@5 = \sum_{i=1}^{5} \frac{rel_i}{\log_2(i+1)}$ -- the log denominator grows with rank, so a highly relevant document contributes much more at rank 1 than rank 5. Dividing by the *ideal* DCG (a perfect ranking's score) normalizes this to 0-1, comparable across queries with different numbers of relevant documents.

**Why every A/B test in this notebook uses this exact function**, rather than a simplified proxy: consistency matters more than convenience here -- comparing two configurations only means something if both are scored identically.

In [6]:
qrels = defaultdict(dict)
for _, r in qrels_train_df.iterrows():
    qrels[r[REL_QUERY_ID_COL]][r[REL_DOC_ID_COL]] = float(r[RELEVANCE_COL])

train_queries = train_queries_df[train_queries_df[QUERY_ID_COL].isin(qrels.keys())].drop_duplicates(subset=[QUERY_ID_COL]).reset_index(drop=True)
print(f"Train queries with judgements: {len(train_queries)}")

def dcg_at_k(rels, k=5):
    rels = np.asarray(rels)[:k]
    return np.sum(rels / np.log2(np.arange(2, rels.size + 2))) if rels.size else 0.0

def ndcg_at_k(ranked_ids, rel_map, k=5):
    gains = [rel_map.get(d, 0.0) for d in ranked_ids[:k]]
    idcg = dcg_at_k(sorted(rel_map.values(), reverse=True)[:k], k)
    return dcg_at_k(gains, k) / idcg if idcg > 0 else 0.0

def evaluate(rank_fn_row, queries_df, k=5, desc="model"):
    scores = []
    for _, row in tqdm(queries_df.iterrows(), total=len(queries_df), desc=f"Eval {desc}", leave=False):
        qid = row[QUERY_ID_COL]
        if qid not in qrels or not qrels[qid]:
            continue
        scores.append(ndcg_at_k(rank_fn_row(row, k), qrels[qid], k=k))
    return float(np.mean(scores)) if scores else 0.0


Train queries with judgements: 308


## Ablation 1 (resolves open question #1): does title-doubling help DENSE retrieval?
Never tested in isolation before -- it was carried over from the lexical pipeline by assumption. The mechanism that makes it help BM25 (directly increasing a term-frequency count) doesn't obviously apply to a bi-encoder, which pools the whole text into one fixed-size vector -- repeating text there could dilute the signal just as easily as boost it. Testing rather than assuming.

In [7]:
embedder = SentenceTransformer("BAAI/bge-small-en-v1.5")
BGE_PREFIX = "Represent this sentence for searching relevant passages: "

dense_embeddings = {}
for variant in ["single", "double"]:
    dense_embeddings[variant] = embedder.encode(dense_doc_texts[variant], batch_size=64, show_progress_bar=False, normalize_embeddings=True)
    print(f"Encoded dense doc embeddings: title-{variant}")

def dense_rank(doc_variant, query_clean, k=100, return_scores=False):
    q_emb = embedder.encode([BGE_PREFIX + query_clean], normalize_embeddings=True)
    sims = (dense_embeddings[doc_variant] @ q_emb.T).flatten()
    top_idx = sims.argsort()[::-1][:k]
    ranked = [doc_ids[i] for i in top_idx]
    return (ranked, sims[top_idx]) if return_scores else ranked

title_ablation_rows = []
for variant in ["single", "double"]:
    score = evaluate(lambda row, k, v=variant: dense_rank(v, row["q_dense_raw"], k=k), train_queries, desc=f"Dense title-{variant}")
    title_ablation_rows.append({"title_doubling": variant, "nDCG@5": score})

title_ablation = pd.DataFrame(title_ablation_rows).sort_values("nDCG@5", ascending=False).reset_index(drop=True)
title_ablation


modules.json:   0%|          | 0.00/349 [00:00<?, ?B/s]

config_sentence_transformers.json:   0%|          | 0.00/124 [00:00<?, ?B/s]

README.md: 0.00B [00:00, ?B/s]

sentence_bert_config.json:   0%|          | 0.00/52.0 [00:00<?, ?B/s]

config.json:   0%|          | 0.00/743 [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/133M [00:00<?, ?B/s]

Loading weights:   0%|          | 0/199 [00:00<?, ?it/s]

BertModel LOAD REPORT from: BAAI/bge-small-en-v1.5
Key                     | Status     |  | 
------------------------+------------+--+-
embeddings.position_ids | UNEXPECTED |  | 

Notes:
- UNEXPECTED	:can be ignored when loading from different task/architecture; not ok if you expect identical arch.


tokenizer_config.json:   0%|          | 0.00/366 [00:00<?, ?B/s]

vocab.txt: 0.00B [00:00, ?B/s]

tokenizer.json: 0.00B [00:00, ?B/s]

special_tokens_map.json:   0%|          | 0.00/125 [00:00<?, ?B/s]

config.json:   0%|          | 0.00/190 [00:00<?, ?B/s]

Encoded dense doc embeddings: title-single
Encoded dense doc embeddings: title-double


Eval Dense title-single:   0%|          | 0/308 [00:00<?, ?it/s]

Eval Dense title-double:   0%|          | 0/308 [00:00<?, ?it/s]

,title_doubling,nDCG@5
0,double,0.768763
1,single,0.763919


In [8]:
BEST_TITLE_VARIANT = title_ablation.iloc[0]["title_doubling"]
print(f"Best doc-text variant for dense: title-{BEST_TITLE_VARIANT} -> nDCG@5={title_ablation.iloc[0]['nDCG@5']:.4f}")
print(f"(v6 used title-double by default and got 0.7495 -- {'confirmed as the right choice' if BEST_TITLE_VARIANT=='double' else 'this was actually costing accuracy'})")


Best doc-text variant for dense: title-double -> nDCG@5=0.7688
(v6 used title-double by default and got 0.7495 -- confirmed as the right choice)


## Ablation 2 (resolves open question #1, part 2): does query expansion help DENSE retrieval?
Tested using whichever doc-text variant won above. This is the more likely of the two suspects for the earlier dense-score regression (0.7688 -> 0.7495): the expansion dictionary was designed to bridge *lexical* vocabulary gaps, and was never validated for a model that already handles synonyms semantically.

In [9]:
dense_expansion_rows = []
for qvar in ["raw", "expanded"]:
    qcol = f"q_dense_{qvar}"
    score = evaluate(lambda row, k, c=qcol: dense_rank(BEST_TITLE_VARIANT, row[c], k=k), train_queries, desc=f"Dense query-{qvar}")
    dense_expansion_rows.append({"query": qvar, "nDCG@5": score})

dense_expansion_ablation = pd.DataFrame(dense_expansion_rows).sort_values("nDCG@5", ascending=False).reset_index(drop=True)
dense_expansion_ablation


Eval Dense query-raw:   0%|          | 0/308 [00:00<?, ?it/s]

Eval Dense query-expanded:   0%|          | 0/308 [00:00<?, ?it/s]

,query,nDCG@5
0,raw,0.768763
1,expanded,0.749541


In [10]:
BEST_DENSE_QUERY_VARIANT = dense_expansion_ablation.iloc[0]["query"]
dense_ndcg = dense_expansion_ablation.iloc[0]["nDCG@5"]
print(f"Best query variant for dense: '{BEST_DENSE_QUERY_VARIANT}' -> nDCG@5={dense_ndcg:.4f}")
print(f"Comparison: v6 dense (double title + expanded query) = 0.7495 | earlier pre-lemmatization run (single title + raw query) = ~0.769")
if dense_ndcg > 0.755:
    print(">>> Isolating these two variables recovered most or all of the earlier 0.769 dense score.")


Best query variant for dense: 'raw' -> nDCG@5=0.7688
Comparison: v6 dense (double title + expanded query) = 0.7495 | earlier pre-lemmatization run (single title + raw query) = ~0.769
>>> Isolating these two variables recovered most or all of the earlier 0.769 dense score.


## Ablation 3 (resolves open question #2): stemming vs lemmatization for BM25/TF-IDF
Crossed with raw vs expanded query -- 4 combinations each for TF-IDF and BM25, 8 evaluations total.

**Why this matters beyond picking a winner:** stemming (Porter) is a crude, rule-based suffix-stripper -- fast, but can over-truncate distinct words to the same root. Lemmatization uses a POS tagger plus a dictionary (WordNet) to return the true base form, which is more linguistically correct but slower and depends on the POS tagger being accurate on short, jargon-heavy factsheet titles it wasn't necessarily tuned for. Neither is obviously better in the abstract for this specific domain -- hence testing both directly.

In [11]:
tfidf_objs, bm25_objs = {}, {}
for mode in ["stem", "lemma"]:
    tfidf = TfidfVectorizer(max_features=50000, ngram_range=(1, 2))
    doc_tfidf = tfidf.fit_transform(lex_doc_texts[mode])
    tfidf_objs[mode] = (tfidf, doc_tfidf)
    bm25_objs[mode] = BM25Okapi([t.split() for t in lex_doc_texts[mode]], k1=1.5, b=0.75)

def tfidf_rank(mode, query_clean, k=5):
    tfidf, doc_tfidf = tfidf_objs[mode]
    sims = cosine_similarity(tfidf.transform([query_clean]), doc_tfidf).flatten()
    return [doc_ids[i] for i in sims.argsort()[::-1][:k]]

def bm25_rank(mode, query_clean, k=100, return_scores=False):
    scores = bm25_objs[mode].get_scores(query_clean.split())
    top_idx = scores.argsort()[::-1][:k]
    ranked = [doc_ids[i] for i in top_idx]
    return (ranked, scores[top_idx]) if return_scores else ranked

lexical_ablation_rows = []
for mode in ["stem", "lemma"]:
    for qvar in ["raw", "expanded"]:
        qcol = f"q_lex_{mode}_{qvar}"
        t_score = evaluate(lambda row, k, c=qcol: tfidf_rank(mode, row[c], k), train_queries, desc=f"TF-IDF {mode}/{qvar}")
        b_score = evaluate(lambda row, k, c=qcol: bm25_rank(mode, row[c], k=5), train_queries, desc=f"BM25 {mode}/{qvar}")
        lexical_ablation_rows.append({"method": "TF-IDF", "morphology": mode, "query": qvar, "nDCG@5": t_score})
        lexical_ablation_rows.append({"method": "BM25", "morphology": mode, "query": qvar, "nDCG@5": b_score})

lexical_ablation = pd.DataFrame(lexical_ablation_rows).sort_values("nDCG@5", ascending=False).reset_index(drop=True)
lexical_ablation


Eval TF-IDF stem/raw:   0%|          | 0/308 [00:00<?, ?it/s]

Eval BM25 stem/raw:   0%|          | 0/308 [00:00<?, ?it/s]

Eval TF-IDF stem/expanded:   0%|          | 0/308 [00:00<?, ?it/s]

Eval BM25 stem/expanded:   0%|          | 0/308 [00:00<?, ?it/s]

Eval TF-IDF lemma/raw:   0%|          | 0/308 [00:00<?, ?it/s]

Eval BM25 lemma/raw:   0%|          | 0/308 [00:00<?, ?it/s]

Eval TF-IDF lemma/expanded:   0%|          | 0/308 [00:00<?, ?it/s]

Eval BM25 lemma/expanded:   0%|          | 0/308 [00:00<?, ?it/s]

,method,morphology,query,nDCG@5
0,TF-IDF,lemma,raw,0.588613
1,TF-IDF,stem,raw,0.582529
2,TF-IDF,lemma,expanded,0.581739
3,TF-IDF,stem,expanded,0.578908
4,BM25,stem,raw,0.511705
5,BM25,lemma,raw,0.499813
6,BM25,stem,expanded,0.497148
7,BM25,lemma,expanded,0.487438


In [12]:
best_lex_row = lexical_ablation.iloc[0]
BEST_MORPH = best_lex_row["morphology"]
BEST_LEX_QUERY_VARIANT = best_lex_row["query"]
print(f"Best lexical config: {best_lex_row['method']} + {BEST_MORPH} + {BEST_LEX_QUERY_VARIANT}-query -> nDCG@5={best_lex_row['nDCG@5']:.4f}")


Best lexical config: TF-IDF + lemma + raw-query -> nDCG@5=0.5886


## BM25 k1/b grid search, using the winning morphology + query variant

**What these parameters mean, briefly:** $k_1$ controls how quickly repeated occurrences of a term stop adding value (diminishing returns -- the 10th "nitrogen" barely matters more than the 5th), and $b$ controls how much a document's length is penalized relative to the corpus average. Default values (`k1=1.5, b=0.75`) are common starting points, not guaranteed optimal for this specific 695-document corpus -- grid search finds the actual best combination rather than trusting the defaults.

In [13]:
BM25_QCOL = f"q_lex_{BEST_MORPH}_{BEST_LEX_QUERY_VARIANT}"
tokenized_docs = [t.split() for t in lex_doc_texts[BEST_MORPH]]

bm25_default_ndcg = evaluate(lambda row, k: bm25_rank(BEST_MORPH, row[BM25_QCOL], k=5), train_queries, desc="BM25 default")
best_bm25 = (bm25_default_ndcg, 1.5, 0.75)
for k1 in [1.2, 1.5, 1.8, 2.0]:
    for b in [0.0, 0.25, 0.5, 0.75]:
        tmp = BM25Okapi(tokenized_docs, k1=k1, b=b)
        def _r(q, k=5, _bm=tmp):
            s = _bm.get_scores(q.split())
            return [doc_ids[i] for i in s.argsort()[::-1][:k]]
        sc = evaluate(lambda row, k, _rf=_r: _rf(row[BM25_QCOL], k), train_queries, desc=f"BM25 k1={k1} b={b}")
        if sc > best_bm25[0]:
            best_bm25 = (sc, k1, b)

bm25_ndcg, best_k1, best_b = best_bm25
bm25_objs[BEST_MORPH] = BM25Okapi(tokenized_docs, k1=best_k1, b=best_b)
print(f"Best BM25: nDCG@5={bm25_ndcg:.4f}  k1={best_k1}  b={best_b}  (default was {bm25_default_ndcg:.4f})")


Eval BM25 default:   0%|          | 0/308 [00:00<?, ?it/s]

Eval BM25 k1=1.2 b=0.0:   0%|          | 0/308 [00:00<?, ?it/s]

Eval BM25 k1=1.2 b=0.25:   0%|          | 0/308 [00:00<?, ?it/s]

Eval BM25 k1=1.2 b=0.5:   0%|          | 0/308 [00:00<?, ?it/s]

Eval BM25 k1=1.2 b=0.75:   0%|          | 0/308 [00:00<?, ?it/s]

Eval BM25 k1=1.5 b=0.0:   0%|          | 0/308 [00:00<?, ?it/s]

Eval BM25 k1=1.5 b=0.25:   0%|          | 0/308 [00:00<?, ?it/s]

Eval BM25 k1=1.5 b=0.5:   0%|          | 0/308 [00:00<?, ?it/s]

Eval BM25 k1=1.5 b=0.75:   0%|          | 0/308 [00:00<?, ?it/s]

Eval BM25 k1=1.8 b=0.0:   0%|          | 0/308 [00:00<?, ?it/s]

Eval BM25 k1=1.8 b=0.25:   0%|          | 0/308 [00:00<?, ?it/s]

Eval BM25 k1=1.8 b=0.5:   0%|          | 0/308 [00:00<?, ?it/s]

Eval BM25 k1=1.8 b=0.75:   0%|          | 0/308 [00:00<?, ?it/s]

Eval BM25 k1=2.0 b=0.0:   0%|          | 0/308 [00:00<?, ?it/s]

Eval BM25 k1=2.0 b=0.25:   0%|          | 0/308 [00:00<?, ?it/s]

Eval BM25 k1=2.0 b=0.5:   0%|          | 0/308 [00:00<?, ?it/s]

Eval BM25 k1=2.0 b=0.75:   0%|          | 0/308 [00:00<?, ?it/s]

Best BM25: nDCG@5=0.5343  k1=2.0  b=0.25  (default was 0.4998)


## Convex combination fusion, using the winning config from every ablation above

**Why convex combination instead of Reciprocal Rank Fusion (RRF):** RRF scores a document only by its *rank position* in each retriever's list ($\frac{1}{k + rank_r(d)}$), never by the retriever's actual confidence. This means RRF can't express "trust dense much more than BM25 here" -- it applies the same fixed rank-based weight regardless of how strong or weak each signal actually is. Measured directly in an earlier run: RRF fusion (0.640) scored *worse* than dense retrieval alone (0.769), because a weak lexical signal was dragging a strong dense signal down at a fixed rate.

Convex combination instead blends the actual **normalized scores** directly ($\alpha \cdot S_{BM25} + (1-\alpha) \cdot S_{dense}$), preserving the *size* of the confidence gap between a clear top match and a crowded tie further down. Since BM25 scores are unbounded and dense cosine similarities are bounded in [-1,1], both are min-max normalized to [0,1] before combining. Alpha is swept empirically rather than assumed -- alpha=0 would mean BM25 contributes nothing to the final ranking, which is itself a valid, informative outcome if that's what the data shows.

In [14]:
POOL = 60
DENSE_QCOL = f"q_dense_{BEST_DENSE_QUERY_VARIANT}"

def minmax_normalize(scores):
    s = np.asarray(scores, dtype=float)
    return np.zeros_like(s) if s.max() == s.min() else (s - s.min()) / (s.max() - s.min())

def convex_fusion_rank(bm25_query, dense_query, k=5, alpha=0.3, pool=POOL):
    bm25_docs, bm25_scores = bm25_rank(BEST_MORPH, bm25_query, k=pool, return_scores=True)
    dense_docs, dense_scores = dense_rank(BEST_TITLE_VARIANT, dense_query, k=pool, return_scores=True)
    bm25_n, dense_n = minmax_normalize(bm25_scores), minmax_normalize(dense_scores)
    bm_dict, den_dict = dict(zip(bm25_docs, bm25_n)), dict(zip(dense_docs, dense_n))
    all_docs = set(bm25_docs) | set(dense_docs)
    combined = {d: alpha * bm_dict.get(d, 0.0) + (1 - alpha) * den_dict.get(d, 0.0) for d in all_docs}
    return sorted(combined, key=combined.get, reverse=True)[:k]

best_alpha, best_alpha_score = 0.3, 0.0
for alpha in [0.0, 0.1, 0.2, 0.3, 0.4, 0.5, 0.6, 0.7, 0.8]:
    sc = evaluate(
        lambda row, k, a=alpha: convex_fusion_rank(row[BM25_QCOL], row[DENSE_QCOL], k, alpha=a),
        train_queries, desc=f"Fusion alpha={alpha}"
    )
    if sc > best_alpha_score:
        best_alpha, best_alpha_score = alpha, sc

print(f"Best alpha={best_alpha} -> nDCG@5={best_alpha_score:.4f}  (vs dense-alone {dense_ndcg:.4f})")


Eval Fusion alpha=0.0:   0%|          | 0/308 [00:00<?, ?it/s]

Eval Fusion alpha=0.1:   0%|          | 0/308 [00:00<?, ?it/s]

Eval Fusion alpha=0.2:   0%|          | 0/308 [00:00<?, ?it/s]

Eval Fusion alpha=0.3:   0%|          | 0/308 [00:00<?, ?it/s]

Eval Fusion alpha=0.4:   0%|          | 0/308 [00:00<?, ?it/s]

Eval Fusion alpha=0.5:   0%|          | 0/308 [00:00<?, ?it/s]

Eval Fusion alpha=0.6:   0%|          | 0/308 [00:00<?, ?it/s]

Eval Fusion alpha=0.7:   0%|          | 0/308 [00:00<?, ?it/s]

Eval Fusion alpha=0.8:   0%|          | 0/308 [00:00<?, ?it/s]

Best alpha=0.0 -> nDCG@5=0.7688  (vs dense-alone 0.7688)


## Cross-encoder reranking: ms-marco-MiniLM-L-6-v2 (the validated choice)

**Why a second model after fusion already produces a ranking:** bi-encoders (like the dense model above) encode the query and each document *completely independently*, then compare with a simple dot product -- fast, since embeddings can be precomputed once, but the model never lets query and document words directly interact while forming a judgment. A cross-encoder instead feeds the query and a candidate document into one transformer input together, so every query token can attend to every document token at every layer -- catching subtle distinctions (negations, which of two near-identical intent words actually matches) that two independently-computed vectors structurally cannot represent. The cost is one full forward pass *per candidate*, which is why this only runs on the ~60-candidate shortlist from fusion, not the full 695-document corpus.

**Not re-testing `bge-reranker-base` here** -- an earlier run already confirmed it loses to MiniLM on this domain, zero-shot, at ~12x more compute. Re-litigating a settled, evidence-backed comparison would just waste runtime.

In [15]:
cross_encoder = CrossEncoder("cross-encoder/ms-marco-MiniLM-L-6-v2")
RERANK_POOL = 60
doc_text_by_id_dense = dict(zip(doc_ids, dense_doc_texts[BEST_TITLE_VARIANT]))

def rerank_rank(row, k=5, pool=RERANK_POOL, alpha=None):
    a = alpha if alpha is not None else best_alpha
    candidates = convex_fusion_rank(row[BM25_QCOL], row[DENSE_QCOL], k=pool, alpha=a)
    if not candidates:
        return []
    pairs = [[row[DENSE_QCOL], doc_text_by_id_dense[d]] for d in candidates]
    ce_scores = cross_encoder.predict(pairs, batch_size=32)
    order = np.argsort(ce_scores)[::-1]
    return [candidates[i] for i in order][:k]

rerank_ndcg = evaluate(lambda row, k: rerank_rank(row, k), train_queries, desc="Fusion + MiniLM rerank")
print(f"Fusion + MiniLM rerank nDCG@5: {rerank_ndcg:.4f}")
print(f"(v6 got 0.7573 with title-double + expanded-dense-query; this run isolated those two variables)")


config.json:   0%|          | 0.00/794 [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/90.9M [00:00<?, ?B/s]

Loading weights:   0%|          | 0/105 [00:00<?, ?it/s]

BertForSequenceClassification LOAD REPORT from: cross-encoder/ms-marco-MiniLM-L-6-v2
Key                          | Status     |  | 
-----------------------------+------------+--+-
bert.embeddings.position_ids | UNEXPECTED |  | 

Notes:
- UNEXPECTED	:can be ignored when loading from different task/architecture; not ok if you expect identical arch.


tokenizer_config.json: 0.00B [00:00, ?B/s]

vocab.txt: 0.00B [00:00, ?B/s]

tokenizer.json: 0.00B [00:00, ?B/s]

special_tokens_map.json:   0%|          | 0.00/132 [00:00<?, ?B/s]

Eval Fusion + MiniLM rerank:   0%|          | 0/308 [00:00<?, ?it/s]

Fusion + MiniLM rerank nDCG@5: 0.7821
(v6 got 0.7573 with title-double + expanded-dense-query; this run isolated those two variables)


## Full ablation results summary
Every method actually evaluated on the real train qrels -- not a single cherry-picked number, but the full comparison across every design choice tested above, so the final selection is traceable back to evidence.

In [16]:
results = pd.DataFrame({
    "Method": [
        f"TF-IDF ({BEST_MORPH}, {BEST_LEX_QUERY_VARIANT})",
        f"BM25 ({BEST_MORPH}, {BEST_LEX_QUERY_VARIANT}, tuned)",
        f"Dense (title-{BEST_TITLE_VARIANT}, {BEST_DENSE_QUERY_VARIANT})",
        f"Convex fusion (alpha={best_alpha})",
        "Fusion + MiniLM rerank",
    ],
    "nDCG@5": [
    lexical_ablation[
        (lexical_ablation.method == "TF-IDF") &
        (lexical_ablation.morphology == BEST_MORPH) &
        (lexical_ablation["query"] == BEST_LEX_QUERY_VARIANT)   # <-- bracket notation, not .query
    ]["nDCG@5"].iloc[0],
    bm25_ndcg,
    dense_ndcg,
    best_alpha_score,
    rerank_ndcg,
],
}).sort_values("nDCG@5", ascending=False).reset_index(drop=True)
results


,Method,nDCG@5
0,Fusion + MiniLM rerank,0.782072
1,Convex fusion (alpha=0.0),0.768763
2,"Dense (title-double, raw)",0.768763
3,"TF-IDF (lemma, raw)",0.588613
4,"BM25 (lemma, raw, tuned)",0.534295


## 5-fold cross-validation of the FINAL winning configuration

Hyperparameters (morphology, query variants, alpha, title-doubling, k1/b) were already chosen above on the full 308 train queries -- **this section does NOT re-tune anything.** It re-measures the final, fixed pipeline across 5 different train/held-out splits, reporting a mean +/- standard deviation instead of one number computed on the same 308 queries that were used throughout the tuning process above. A wide spread here would itself be informative: it would mean the single-number results above are less trustworthy than they look, since they'd be sensitive to exactly which queries happened to be in the training sample.

In [17]:
method_fns_final = {
    f"TF-IDF ({BEST_MORPH}, {BEST_LEX_QUERY_VARIANT})": lambda row, k: tfidf_rank(BEST_MORPH, row[BM25_QCOL], k),
    f"BM25 ({BEST_MORPH}, {BEST_LEX_QUERY_VARIANT}, tuned)": lambda row, k: bm25_rank(BEST_MORPH, row[BM25_QCOL], k),
    f"Dense (title-{BEST_TITLE_VARIANT}, {BEST_DENSE_QUERY_VARIANT})": lambda row, k: dense_rank(BEST_TITLE_VARIANT, row[DENSE_QCOL], k),
    f"Convex fusion (alpha={best_alpha})": lambda row, k: convex_fusion_rank(row[BM25_QCOL], row[DENSE_QCOL], k, alpha=best_alpha),
    "Fusion + MiniLM rerank": lambda row, k: rerank_rank(row, k),
}
best_method_name = results.iloc[0]["Method"]
final_rank_fn = method_fns_final[best_method_name]
print(f"Cross-validating: {best_method_name}")

kf = KFold(n_splits=5, shuffle=True, random_state=42)
qid_array = train_queries[QUERY_ID_COL].values
fold_scores = []
for fold_i, (_, val_idx) in enumerate(kf.split(qid_array)):
    val_qids = set(qid_array[val_idx])
    val_queries = train_queries[train_queries[QUERY_ID_COL].isin(val_qids)]
    fold_ndcg = evaluate(lambda row, k: final_rank_fn(row, k), val_queries, desc=f"Fold {fold_i}")
    fold_scores.append(fold_ndcg)
    print(f"  Fold {fold_i}: nDCG@5 = {fold_ndcg:.4f}  (n={len(val_queries)})")

fold_scores = np.array(fold_scores)
print(f"\n5-fold CV: mean={fold_scores.mean():.4f}  std={fold_scores.std():.4f}")
print(f"Full-308 single-number result was: {results.iloc[0]['nDCG@5']:.4f}")
print(f"(Previous best public LB score: 0.83823)")


Cross-validating: Fusion + MiniLM rerank


Eval Fold 0:   0%|          | 0/62 [00:00<?, ?it/s]

  Fold 0: nDCG@5 = 0.7759  (n=62)


Eval Fold 1:   0%|          | 0/62 [00:00<?, ?it/s]

  Fold 1: nDCG@5 = 0.8200  (n=62)


Eval Fold 2:   0%|          | 0/62 [00:00<?, ?it/s]

  Fold 2: nDCG@5 = 0.7847  (n=62)


Eval Fold 3:   0%|          | 0/61 [00:00<?, ?it/s]

  Fold 3: nDCG@5 = 0.7836  (n=61)


Eval Fold 4:   0%|          | 0/61 [00:00<?, ?it/s]

  Fold 4: nDCG@5 = 0.7456  (n=61)

5-fold CV: mean=0.7820  std=0.0237
Full-308 single-number result was: 0.7821
(Previous best public LB score: 0.83823)


## Generate submission.csv in the REQUIRED long format
Header `QueryId,DocumentId`, exactly 5 rows per test query, row order = predicted rank (best first), no duplicate (QueryId, DocumentId) pairs -- these constraints are validated explicitly in the code below rather than assumed to hold.

In [18]:
print(f"Using: {best_method_name} (full-308 local nDCG@5 = {results.iloc[0]['nDCG@5']:.4f}, 5-fold CV mean = {fold_scores.mean():.4f} +/- {fold_scores.std():.4f})")

rows = []
for _, row in tqdm(test_queries_df.iterrows(), total=len(test_queries_df), desc="Predicting test queries"):
    qid = row[QUERY_ID_COL]
    ranked = final_rank_fn(row, 5)
    seen = set()
    ranked = [d for d in ranked if not (d in seen or seen.add(d))]
    fallback_iter = iter(doc_ids)
    while len(ranked) < 5:
        d = next(fallback_iter)
        if d not in ranked:
            ranked.append(d)
    for d in ranked[:5]:
        rows.append({"QueryId": qid, "DocumentId": d})

submission_df = pd.DataFrame(rows)
print(f"\nRows: {len(submission_df)}  (expect {len(test_queries_df)*5})")
print(f"Rows per query all == 5: {(submission_df.groupby('QueryId').size() == 5).all()}")
print(f"No duplicate (QueryId, DocumentId): {not submission_df.duplicated(subset=['QueryId','DocumentId']).any()}")
submission_df.head(10)


Using: Fusion + MiniLM rerank (full-308 local nDCG@5 = 0.7821, 5-fold CV mean = 0.7820 +/- 0.0237)


Predicting test queries:   0%|          | 0/200 [00:00<?, ?it/s]


Rows: 1000  (expect 1000)
Rows per query all == 5: True
No duplicate (QueryId, DocumentId): True


,QueryId,DocumentId
0,1001,1
1,1001,4
2,1001,3
3,1001,5
4,1001,2
5,1002,4
6,1002,3
7,1002,5
8,1002,2
9,1002,1


In [19]:
os.makedirs("/kaggle/working", exist_ok=True)
submission_df.to_csv("/kaggle/working/submission.csv", index=False)
print("Saved to /kaggle/working/submission.csv")
pd.read_csv("/kaggle/working/submission.csv").head()


Saved to /kaggle/working/submission.csv


,QueryId,DocumentId
0,1001,1
1,1001,4
2,1001,3
3,1001,5
4,1001,2


## Per-query error analysis on the final method
Sorts train queries by nDCG@5 ascending, so the worst failures can be inspected directly -- this is the same process that originally uncovered the stemming fix (a near-duplicate document cluster where the one discriminating title word wasn't matching an unstemmed query). Aggregate metrics hide *why* a system fails; reading the actual retrieved-vs-gold documents for the worst cases is what turns a failure into an actionable fix.

In [20]:
per_query_scores = []
for _, row in train_queries.iterrows():
    qid = row[QUERY_ID_COL]
    if qid not in qrels or not qrels[qid]:
        continue
    ranked = final_rank_fn(row, 5)
    score = ndcg_at_k(ranked, qrels[qid])
    per_query_scores.append({
        "query_id": qid, "query": row[QUERY_TEXT_COL], "nDCG@5": score,
        "retrieved": ranked, "gold_top": sorted(qrels[qid], key=qrels[qid].get, reverse=True)[:5]
    })

per_query_df = pd.DataFrame(per_query_scores).sort_values("nDCG@5").reset_index(drop=True)
print("Worst 10 train queries under the final method:")
per_query_df.head(10)


Worst 10 train queries under the final method:


,query_id,query,nDCG@5,retrieved,gold_top
0,308,What should I do about an outbreak of soil acidification in maize?,0.000000,"[115, 131, 99, 123, 601]","[687.0, 690.0, 693.0, 686.0, 223.0]"
1,307,How do I manage soil acidification in maize?,0.000000,"[693, 115, 99, 131, 222]","[687.0, 690.0, 693.0, 686.0, 223.0]"
2,306,What should I do about an outbreak of nutrient deficiency in maize?,0.000000,"[348, 350, 347, 346, 428]","[686.0, 694.0, 693.0, 687.0, 192.0]"
3,305,How do I manage nutrient deficiency in maize?,0.000000,"[350, 348, 346, 430, 349]","[686.0, 694.0, 693.0, 687.0, 192.0]"
4,57,How do I manage rosette disease in groundnut?,0.000000,"[105, 107, 103, 106, 108]","[191.0, 101.0, 102.0, 104.0, 106.0]"
5,58,What should I do about an outbreak of rosette disease in groundnut?,0.000000,"[105, 107, 103, 108, 106]","[191.0, 101.0, 102.0, 104.0, 106.0]"
6,21,How do I manage blast in rice?,0.000000,"[139, 137, 135, 140, 150]","[59.0, 150.0, 152.0, 149.0, 134.0]"
7,22,What should I do about an outbreak of blast in rice?,0.386853,"[139, 137, 135, 140, 59]","[59.0, 150.0, 152.0, 149.0, 134.0]"
8,283,How do I control cowpea insect pests on cowpea?,0.386853,"[606, 608, 652, 604, 645]","[645.0, 652.0, 653.0, 651.0, 647.0]"
9,113,How do I fix nitrogen deficiency in rice?,0.387227,"[371, 373, 370, 372, 369]","[368.0, 370.0, 372.0, 336.0, 333.0]"


## Summary of what this run established
- Whether title-doubling helps or hurts dense retrieval -- previously assumed, now measured directly (it helps: 0.7688 vs 0.7639).
- Whether query expansion helps or hurts dense retrieval -- previously assumed, now measured directly (it hurts dense: 0.7495 vs 0.7688 raw; kept on the lexical side only).
- Stemming vs lemmatization for the lexical path, decided empirically rather than by preference.
- BM25's real contribution to fusion, using genuinely tuned parameters rather than defaults.
- A 5-fold CV estimate of the final pipeline's reliability, not just a single point estimate (mean 0.782, std 0.024).

## Final competition outcome (added after the private leaderboard was released)
This notebook's output was our actual final submission. Once the organizers scored it against the private (held-out) test set: **Public 0.83823, Private 0.86982**, placing **8th overall**.

**The local/leaderboard relationship, resolved:** throughout development, local training-set nDCG@5 did not reliably predict *relative* ranking between pipeline variants -- at one point, a simpler dense-only configuration outscored this more elaborate fusion+rerank pipeline on the public leaderboard despite scoring lower locally. We treated this as an open overfitting concern in earlier notebook versions. With the private score now known (0.86982, *higher* than both the local estimate of 0.782 and the public score of 0.838), the more likely explanation in hindsight is that the 200 public/private test queries and the 308 training queries are simply somewhat different samples of the same underlying distribution -- normal sampling variance across three different query sets (train/public-test/private-test), not systematic overfitting from our tuning process. The 5-fold CV's standard deviation of 0.024 on the training set alone was already a meaningful hint that single-point comparisons this close together shouldn't be over-interpreted.

**Takeaway for future iterations:** the empirical, test-before-adopting methodology used throughout this notebook (stemming, convex fusion, MiniLM over a larger reranker, title-doubling, dropping query expansion from the dense path) was validated by the outcome -- the private score exceeded both our local estimate and the public score, rather than underperforming it, which is the direction you'd hope for rather than the direction that would indicate overfitting to the training or public-test sets specifically.